crates/bcr-wdc-ebpp/src/onchain.rs

```diff
@@ -0,0 +1,314 @@
+// ----- standard library imports
+use std::sync::{Arc, Mutex};
+// ----- extra library imports
+use anyhow::anyhow;
+use async_trait::async_trait;
+use bdk_esplora::EsploraAsyncExt;
+use bdk_wallet::{
+    bitcoin::{bip32::Xpriv, hashes::Hash, Network},
+    descriptor::template::Bip84,
+    keys::{bip39::Mnemonic, DerivableKey, ExtendedKey},
+    miniscript::{descriptor::KeyMap, Descriptor, DescriptorPublicKey},
+    rusqlite::OpenFlags,
+    KeychainKind,
+};
+use futures::future::JoinAll;
+use rand::Rng;
+// ----- local imports
+use crate::error::{Error, Result};
+use crate::service::OnChainWallet;
+
+// ----- end imports
+
+pub type BdkWallet = bdk_wallet::PersistedWallet<bdk_wallet::rusqlite::Connection>;
+pub type PersistedBdkWallet = (BdkWallet, bdk_wallet::rusqlite::Connection);
+pub type SingleKeyWallet = (Descriptor<DescriptorPublicKey>, KeyMap);
+
+#[derive(Debug, Clone, serde::Deserialize)]
+pub struct WalletConfig {
+    mnemonic: Mnemonic,
+    network: Network,
+    store_path: std::path::PathBuf,
+    stop_gap: usize,
+    update_interval: chrono::Duration,
+}
+
+#[async_trait]
+pub trait PrivateKeysRepository {
+    async fn get_private_keys(&self) -> Result<Vec<SingleKeyWallet>>;
+    async fn add_key(&self, key: SingleKeyWallet) -> Result<()>;
+}
+
+#[derive(Debug, Clone)]
+pub struct Wallet<KeyRepo, Syncer> {
+    main: Arc<Mutex<PersistedBdkWallet>>,
+    onetimes: Arc<Mutex<Vec<Arc<Mutex<PersistedBdkWallet>>>>>,
```

Comment:
For a type like this, it might be nice to have a comment explaining why it's so convoluted (onetimes needs to be mutable by requests, inner wallets are accessed in the wallet update loop)